contrib/mesos/pkg/executor/watcher.go

```diff
@@ -0,0 +1,150 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package executor
+
+import (
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+)
+
+type (
+	// filter registration events, return false to abort further processing of the event
+	watchFilter func(pod *RegisteredPod) (accept bool)
+
+	watchExpiration struct {
+		// timeout closes when the handler has expired; it delivers at most one Time.
+		timeout <-chan time.Time
+
+		// onEvent is an optional callback that is invoked if/when the expired chan
+		// closes
+		onEvent func(taskID string)
+	}
+
+	watchHandler struct {
+		// prevent callbacks from being invoked simultaneously
+		sync.Mutex
+
+		// handle registration events, return true to indicate the handler should be
+		// de-registered upon completion. If pod is nil then the associated handler
+		// has expired.
+		onEvent func(pod *RegisteredPod) (done bool, err error)
+
+		// expiration is an optional configuration that indicates when a handler should
+		// be considered to have expired, and what action to take upon such
+		expiration watchExpiration
+	}
+
+	// watcher observes RegisteredPod events and conditionally executes handlers that
+	// have been associated with the taskID of the RegisteredPod.
+	watcher struct {
+		updates  <-chan *RegisteredPod
+		rw       sync.RWMutex
+		handlers map[string]watchHandler
+		filters  []watchFilter
+		runOnce  chan struct{}
+	}
+)
+
+func newWatcher(updates <-chan *RegisteredPod) *watcher {
+	return &watcher{
+		updates:  updates,
+		handlers: make(map[string]watchHandler),
+		runOnce:  make(chan struct{}),
+	}
+}
+
+func (pw *watcher) run() {
+	select {
+	case <-pw.runOnce:
+		log.Error("run() has already been invoked for this pod-watcher")
+		return
+	default:
+		close(pw.runOnce)
+	}
+updateLoop:
+	for u := range pw.updates {
+		log.V(1).Infof("filtering task %v pod %v/%v", u.taskID, u.Namespace, u.Name)
+		for _, f := range pw.filters {
+			if !f(u) {
+				continue updateLoop
+			}
+		}
+		log.V(1).Infof("handling task %v pod %v/%v", u.taskID, u.Namespace, u.Name)
+		h, ok := func() (h watchHandler, ok bool) {
+			pw.rw.RLock()
+			defer pw.rw.RUnlock()
+			h, ok = pw.handlers[u.taskID]
+			return
+		}()
+		if ok {
+			log.V(1).Infof("executing action for task %v pod %v/%v", u.taskID, u.Namespace, u.Name)
+			done, err := func() (bool, error) {
+				h.Lock()
+				defer h.Unlock()
+				return h.onEvent(u)
+			}()
+			if err != nil {
+				log.Error(err)
+			}
+			if done {
+				// de-register handler upon successful completion of action
+				log.V(1).Infof("de-registering handler for task %v pod %v/%v", u.taskID, u.Namespace, u.Name)
+				func() {
+					pw.rw.Lock()
+					delete(pw.handlers, u.taskID)
+					pw.rw.Unlock()
+				}()
+			}
+		}
+	}
+}
+
+func (pw *watcher) filter(f watchFilter) {
```

Comment:
`addFilter` is a better name IMO. `filter` implies that calling this does the actually filtering.